### PR TITLE
Validate deprecated fields via validation visitor

### LINKED ISSUE #8092: Unify field deprecation and unknown field handling behind the validation visitor
Today, unknown field handling is actioned via the validation visitor but deprecated field handling is done inline. Ideally we have consistency here. This would be a nice cleanup, but isn't hugely impacting in any way beyond OCD.

### PR DESCRIPTION
**Pull request by nezdolik**
Description: Duplicate of #10307
This change intends to unify field deprecation and unknown field handling behind the validation visitor. So far validation of unknown fields was done through validation visitor and validation of deprecated fields was performed inline.
Risk Level:Low
Testing: TBD
Docs Changes:NA
Release Notes:TBD
Fixes #8092

### CODE REVIEW

File: bazel/envoy_internal.bzl [hunk 1]
```diff
@@ -7,7 +7,7 @@ def envoy_copts(repository, test = False):
     posix_options = [
         "-Wall",
         "-Wextra",
-        "-Werror",
+        #        "-Werror",
```
Comment (nezdolik) [inline bazel/envoy_internal.bzl:10]: oops

**Comment by nezdolik** [comment]
still need to update/add more tests

File: source/common/protobuf/utility.cc [hunk 2]
```diff
@@ -197,12 +198,7 @@ void deprecatedFieldHelper(Runtime::Loader* runtime, bool proto_annotated_as_dep
   if (warn_only) {
     ENVOY_LOG_MISC(warn, "{}", with_overridden);
   } else {
-    const char fatal_error[] =
-        " If continued use of this field is absolutely necessary, see "
-        "https://www.envoyproxy.io/docs/envoy/latest/configuration/operations/runtime"
-        "#using-runtime-overrides-for-deprecated-features for how to apply a temporary and "
-        "highly discouraged override.";
-    throw ProtoValidationException(with_overridden + fatal_error, message);
```
Comment (htuch) [inline source/common/protobuf/utility.cc:205]: The validation visitors above take this exception and decide to warn/throw depending on whether they are strict/warning visitors. The validation visitors are configured on CLI to reflect the preferences wrt unknown field handling. I think in both cases, for deprecated, we want them to throw, since the code path that led us here indicates that we must do so.

CC @alyssawilk

Comment (nezdolik) [reply source/common/protobuf/utility.cc:205]: @htuch since the action (log, throw exception) that validator visitor needs to take does not depend on runtime type but on code logic, i could think of multiple (not pretty) approaches how it could be handled:
1. No need to pass visitor into method, invoke static method of corresponding visitor type (`StrictValidationVisitorImpl::onDeprecatedField`) , which involves changing signature of method to static.
2. Share same implementation between 2 visitor types for `onDeprecatedField` method, the implementation would throw an exception. This would be slightly semantically misleading, as warning visitor would throw an exception.

Comment (nezdolik) [reply source/common/protobuf/utility.cc:205]: Unless you have other suggestion, second one sounds like less evil.

Comment (htuch) [reply source/common/protobuf/utility.cc:205]: I think (2) sounds good. I think if we were being pedantic, we would rename the visitors, the warning visitor would be the `UnknownFieldWarningVisitor`, etc.

File: source/common/protobuf/message_validator_impl.h [hunk 3]
```diff
@@ -14,32 +14,42 @@ class NullValidationVisitorImpl : public ValidationVisitor {
...
-  Stats::Counter* counter_{};
-  uint64_t prestats_count_{};
+  // This can be late initialized via setUnknownCounter(), enabling the server bootstrap loading
+  // which occurs prior to the initialization of the stats subsystem.
+  Stats::Counter* unknown_counter_{};
+  // This can be late initialized via setDeprecatedCounter(), enabling the server bootstrap loading
+  // which occurs prior to the initialization of the stats subsystem.
+  Stats::Counter* deprecated_counter_{};
+  uint64_t prestats_unknown_count_{};
+  uint64_t prestats_deprecated_count_{};
```
Comment (nezdolik) [inline source/common/protobuf/message_validator_impl.h:45]: adding extra counter for deprecated fields may not be needed, as there is stats counter in Runtime already: `deprecated_feature_use`

Comment (htuch) [reply source/common/protobuf/message_validator_impl.h:45]: Yep, that's fair enough.

File: source/common/protobuf/utility.cc [hunk 4]
```diff
@@ -197,12 +198,7 @@ void deprecatedFieldHelper(Runtime::Loader* runtime, bool proto_annotated_as_dep
...
+    validation_visitor.onDeprecatedField("type " + message.GetTypeName() + " " + with_overridden);
```
Comment (htuch) [inline source/common/protobuf/utility.cc:201]: I think this should always be called, even if warning, but we can add a parameter to control whether it's expected to log or throw an exception (`warn_only`), maybe call it `soft_deprecation`?

File: source/common/protobuf/message_validator_impl.h (hunk 3 above)
Comment (htuch) [reply source/common/protobuf/message_validator_impl.h:45]: True, it's a bit strange that we bump the stat in runtime rather than in the calling code, but I don't think we can undo that at this point as folks are likely relying on it. So, yeah, we can simplify the code here quite a bit as a result.

**Comment by nezdolik** [comment]
keepalive ping, i was busy with another pr

File: include/envoy/protobuf/message_validator.h [hunk 5]
```diff
@@ -39,6 +56,15 @@ class ValidationVisitor {
...
   virtual bool skipValidation() PURE;
+
+  /**
+   * Invoked when deprecated field is encountered.
+   * @param description human readable description of the field.
+   */
+  virtual void onDeprecatedField(absl::string_view description, bool soft_deprecation) PURE;
+
+protected:
+  void onDeprecatedFieldDefault(absl::string_view description, bool soft_deprecation);
```
Comment (nezdolik) [inline include/envoy/protobuf/message_validator.h:67]: @htuch added this method as default impl to be reused by both validation visitors, but failing to compile with:
```include/envoy/protobuf/message_validator.cc:15:15: error: 'warn' was not declared in this scope
     ENVOY_LOG(warn, "Unexpected field: {}",```

Comment (nezdolik) [reply include/envoy/protobuf/message_validator.h:67]: could also just duplicate this method in both derived classes

**Comment by nezdolik** [comment]
```
Caused by: io.grpc.StatusRuntimeException: FAILED_PRECONDITION: there are no bots capable of executing the action, requested action properties: OSFamily = Linux, concurrency = exclusive (Added by RBE; concurrent actions are incompatible with the following requested properties: dockerSiblingContainers = true)
```
Not sure how to fix this

File: test/mocks/protobuf/mocks.h [hunk 6]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include "envoy/protobuf/message_validator.h"
+#include "envoy/stats/stats.h"
```
Comment (nezdolik) [inline test/mocks/protobuf/mocks.h:4]: not needed anymore

File: include/envoy/protobuf/message_validator.cc [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+#include "envoy/common/exception.h"
+
+#include "common/common/logger.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+namespace ProtobufMessage {
+
+void ValidationVisitor::onDeprecatedFieldDefault(absl::string_view description,
```
Comment (htuch) [inline include/envoy/protobuf/message_validator.cc:12]: We tend to prefer that even default behavior lives in `source/` vs. `include/`, the latter of which is generally just pure virtual functions and classes.

File: include/envoy/protobuf/message_validator.h [hunk 8]
```diff
@@ -3,22 +3,39 @@
...
+#include "common/common/documentation_url.h"
 #include "common/protobuf/protobuf.h"
 
 #include "absl/strings/string_view.h"
 
 namespace Envoy {
 namespace ProtobufMessage {
 
+namespace ValidationError {
+const char deprecation_error[] = " If continued use of this field is absolutely necessary, "
```
Comment (htuch) [inline include/envoy/protobuf/message_validator.h:15]: This also belongs in `source/`

File: include/envoy/protobuf/BUILD [hunk 9]
```diff
@@ -11,5 +11,10 @@ envoy_package()
 envoy_cc_library(
     name = "message_validator_interface",
     hdrs = ["message_validator.h"],
-    deps = ["//source/common/protobuf"],
+    deps = [
+        "//source/common/common:documentation_url_lib",
```
Comment (htuch) [inline include/envoy/protobuf/BUILD:15]: Are these changes needed anymore?

File: source/common/protobuf/message_validator_impl.cc [hunk 10]
```diff
@@ -24,20 +24,31 @@ void WarningValidationVisitorImpl::onUnknownField(absl::string_view description)
...
+    ++prestats_unknown_count_;
   } else {
-    counter_->inc();
+    unknown_counter_->inc();
   }
 }
 
+void WarningValidationVisitorImpl::onDeprecatedField(absl::string_view description,
+                                                     bool soft_deprecation) {
+  onDeprecatedFieldDefault(description, soft_deprecation);
```
Comment (htuch) [inline source/common/protobuf/message_validator_impl.cc:39]: Maybe just call it `onDeprecatedFieldCommon` or something at this point (since it's no longer inheritance).

**Comment by nezdolik** [comment]
Thanks for review @htuch. I will need to fix few tests that use MockValdationVisitors

**Comment by nezdolik** [comment]
@htuch this PR is ready for review

File: source/common/protobuf/message_validator_impl.cc [hunk 11]
```diff
@@ -24,12 +24,22 @@ void WarningValidationVisitorImpl::onUnknownField(absl::string_view description)
   if (!it.second) {
     return;
   }
+
   // It's a new field, log and bump stat.
-  ENVOY_LOG(warn, "Unknown field: {}", description);
-  if (counter_ == nullptr) {
-    ++prestats_count_;
+  ENVOY_LOG(warn, "Unexpected field: {}", description);
```
Comment (htuch) [inline source/common/protobuf/message_validator_impl.cc:29]: Why did you change `unknown` to `unexpected` here?

File: source/common/protobuf/message_validator_impl.cc [hunk 12]
```diff
@@ -38,6 +48,15 @@ void StrictValidationVisitorImpl::onUnknownField(absl::string_view description)
       absl::StrCat("Protobuf message (", description, ") has unknown fields"));
 }
 
+void StrictValidationVisitorImpl::onDeprecatedField(absl::string_view description,
+                                                    bool soft_deprecation) {
+  if (soft_deprecation) {
+    ENVOY_LOG_MISC(warn, "Unexpected field: {}", absl::StrCat(description, deprecation_error));
```
Comment (htuch) [inline source/common/protobuf/message_validator_impl.cc:54]: Should this be `Deprecated field: {}`?

File: source/common/protobuf/message_validator_impl.cc [hunk 13]
```diff
@@ -24,12 +24,22 @@ void WarningValidationVisitorImpl::onUnknownField(absl::string_view description)
...
+}
+
+void WarningValidationVisitorImpl::onDeprecatedField(absl::string_view description,
+                                                     bool soft_deprecation) {
+  if (soft_deprecation) {
+    ENVOY_LOG_MISC(warn, "Unexpected field: {}", absl::StrCat(description, deprecation_error));
   } else {
-    counter_->inc();
+    throw DeprecatedProtoFieldException(absl::StrCat(description, deprecation_error));
   }
```
Comment (htuch) [inline source/common/protobuf/message_validator_impl.cc:43]: How come the common method is inline now?

File: source/common/protobuf/message_validator_impl.cc (hunk 11 above)
Comment (nezdolik) [reply source/common/protobuf/message_validator_impl.cc:29]: Looks like copy paste from previous version when `onUnknownField` and `onDeprecatedField` reused same common method.

File: source/common/protobuf/utility.cc [hunk 14]
```diff
@@ -196,14 +197,9 @@ void deprecatedFieldHelper(Runtime::Loader* runtime, bool proto_annotated_as_dep
   std::string with_overridden = fmt::format(
       error,
       (runtime_overridden ? "runtime overrides to continue using now fatal-by-default " : ""));
-  if (warn_only) {
```
Comment (nezdolik) [inline source/common/protobuf/utility.cc:199]: So this code part is now in `onDeprecatedField` method in strict & warning visitors @htuch . Worth to note that new exception type `DeprecatedFieldException` has been introduced instead of `ProtoValidationException`.

File: source/common/protobuf/message_validator_impl.cc (hunk 12 above)
Comment (nezdolik) [reply source/common/protobuf/message_validator_impl.cc:54]: yes

File: source/common/protobuf/message_validator_impl.cc (hunk 13 above)
Comment (nezdolik) [reply source/common/protobuf/message_validator_impl.cc:43]: @htuch if i understand your question correctly, you mean previous version of PR where both visitors shared common method for incrementing stats and logging message? i've changed impl, there is no longer common method. There is little benefit to have common method as deprecated field validation does not increment counters and log messages are different.

Comment (htuch) [reply source/common/protobuf/message_validator_impl.cc:43]: Yeah, but looking at the two implementations, I'm not seeing a stats increment in either, and still unclear why the warnings are different.
